File: 2024/shared/runtime.go
package shared

import (
	"fmt"
	"os"
)

func Run[T any](parse func(string) T, solvePart1 func(input T) any, solvePart2 func(input T) any) {
	example, _ := os.ReadFile("example.txt")
	input, _ := os.ReadFile("input.txt")

	if len(os.Args) < 3 {
		fmt.Println("Usage: go run main.go <example|input> <1|2>")
		os.Exit(1)
	}

	inputType := os.Args[1]
	partNumber := os.Args[2]

	currentInput := string(input)
	if inputType == "example" {
		currentInput = string(example)
	}

	parsed := parse(currentInput)

	if partNumber == "1" {
		fmt.Println(solvePart1(parsed))
	} else {
		fmt.Println(solvePart2(parsed))
	}
}

File: 2024/shared/runtime.ts
export async function run<T>(parse: (input: string) => T, part1: (input: T) => unknown, part2: (input: T) => unknown) {
  const inputType = Deno.args[0];
  const partNumber = Deno.args[1];

  const currentInput = inputType === 'example' ? await readExample() : await readInput();
  const solveFunction = partNumber === '1' ? part1 : part2;

  console.time('Execution time');
  const parsed = parse(currentInput);
  console.log(solveFunction(parsed));
  console.timeEnd('Execution time');
  Deno.exit();
}

export async function readExample() {
  return await Deno.readTextFile('example.txt');
}

export async function readInput() {
  return await Deno.readTextFile('input.txt');
}

export function memoize<T extends (...args: any[]) => any>(func: T): T {
  const cache = new Map<string, ReturnType<T>>();

  return ((...args: Parameters<T>): ReturnType<T> => {
    const key = JSON.stringify(args);
    if (!cache.has(key)) {
      cache.set(key, func(...args));
    }
    return cache.get(key) as ReturnType<T>;
  }) as T;
}
